### PR TITLE
Add options `regex` and `ignore` to rule `macro_names`

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
[Fix inaka/elvis#515]

### CODE REVIEW

File: src/elvis_style.erl [hunk 1]
```diff
@@ -860,24 +877,48 @@ check_no_trailing_whitespace(Line, Num, RuleConfig) ->
...
 %% Macro Names
 
--spec check_macro_names(binary(), integer(), [term()]) ->
+-spec check_macro_names(binary(), integer(), ConfigRegex :: {ok, string()} | error) ->
     no_result | {ok, elvis_result:item()}.
-check_macro_names(Line, Num, _Args) ->
-    {ok, Regex} = re:compile("^ *[-]define *[(]([^,(]+)"),
-    case re:run(Line, Regex, [{capture, all_but_first, list}]) of
+check_macro_names(Line, Num, ConfigRegex) ->
+    {ok, Regex} = re:compile("^ *[-]define *[(] *([']([^']*)[']|[^,( ]+)"),
```
Comment (paulo-ferraz-oliveira) [inline src/elvis_style.erl:883]: I found this regexp to not cover stuff like `-define('A,a', 2).` which, although ugly, might show up in the wild.

Let me know if this change is OK. My goal here is to try and fetch:
1. either `'Macro'` (' followed by a potential repetition of non-' followed by '),
2. or `Macro` (a potential repetition of non ", or (" chars.

File: src/elvis_style.erl [hunk 2]
```diff
@@ -860,24 +877,48 @@ check_no_trailing_whitespace(Line, Num, RuleConfig) ->
...
+    case re:run(Line, Regex, [{capture, [1], list}]) of
```
Comment (paulo-ferraz-oliveira) [inline src/elvis_style.erl:884]: Added this because of the group that's created in the new match.

File: test/style_SUITE.erl [hunk 3]
```diff
@@ -246,7 +246,13 @@ verify_macro_names_rule(_Config) ->
     File = "fail_macro_names.erl",
     {ok, Path} = elvis_test_utils:find_file(SrcDirs, File),
 
-    [_, _] = elvis_style:macro_names(ElvisConfig, Path, #{}).
+    [_, _] = elvis_style:macro_names(ElvisConfig, Path, #{}),
+
+    [_] = elvis_style:macro_names(ElvisConfig, Path, #{ regex => "^[A-Za-z_ ]+$" }),
```
Comment (paulo-ferraz-oliveira) [inline test/style_SUITE.erl:251]: Which a regex stuff that was failing (at least 1) is not failing any more.

File: test/style_SUITE.erl [hunk 4]
```diff
@@ -246,7 +246,13 @@ verify_macro_names_rule(_Config) ->
...
+
+    [] = elvis_style:macro_names(ElvisConfig, Path, #{ regex => "^[A-Za-z_ \-]+$" }),
```
Comment (paulo-ferraz-oliveira) [inline test/style_SUITE.erl:253]: With an "improved" regexp nothing fails.

File: test/style_SUITE.erl [hunk 5]
```diff
@@ -246,7 +246,13 @@ verify_macro_names_rule(_Config) ->
...
+
+    [_, _, _, _, _] = elvis_style:macro_names(ElvisConfig, Path, #{ regex => "^POTENTIAL_BAD-NAME$" }).
```
Comment (paulo-ferraz-oliveira) [inline test/style_SUITE.erl:255]: With a very targeted regexp almost everything fails.

**Comment by paulo-ferraz-oliveira** [comment]
Let me know if I should add more tests or make them more readable.

Actually, now that I think about it I didn't add a test for `ignore`. I'ma do it.

**Comment by paulo-ferraz-oliveira** [comment]
> Let me know if I should add more tests or make them more readable.
> 
> Actually, now that I think about it I didn't add a test for `ignore`. I'ma do it.

There you go: 6057b99.

**Comment by paulo-ferraz-oliveira** [comment]
Added a new regression test for https://github.com/inaka/elvis/issues/515#issuecomment-689870216.

File: src/elvis_style.erl (hunk 1 above)
Comment (elbrujohalcon) [reply src/elvis_style.erl:883]: This is very much *intentional*. We do want to emit a warning if people use `A,a` as a macro name. That's the whole point :)

File: test/examples/fail_macro_names.erl [hunk 6]
```diff
@@ -5,6 +5,8 @@
 -define(wtf_NAME,   "(yuno)").
 -define(GOOD_NAME(Arg), "(megusta)").
 -define(GOOD_NAME_TOO (Arg), "(megusta)").
+-define(  'POTENTIAL_BAD-NAME'  , nomegusta).
+-define('A,aZ', 2).
```
Comment (elbrujohalcon) [inline test/examples/fail_macro_names.erl:9]: These 2 should definitely be rejected by the rule (at least, with default configuration).

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

File: src/elvis_style.erl (hunk 1 above)
Comment (elbrujohalcon) [reply src/elvis_style.erl:883]: In other words: Make the default regex as strict as possible.

Comment (elbrujohalcon) [reply src/elvis_style.erl:883]: Yeah… probably, _the right solution:tm:_ …      would be to actually use the AST for this instead of parsing the code manually as a string.
In any case, this change is fine until that day…

**Review by elbrujohalcon — APPROVED** [review]

File: src/elvis_style.erl [hunk 7]
```diff
@@ -243,13 +247,26 @@ no_trailing_whitespace(_Config, Target, RuleConfig) ->
...
                   elvis_file:file(),
-                  empty_rule_config()) ->
+                  macro_names_config()) ->
     [elvis_result:item()].
-macro_names(_Config, Target, _RuleConfig) ->
+macro_names(Config, Target, RuleConfig) ->
     {Src, _} = elvis_file:src(Target),
-    elvis_utils:check_lines(Src, fun check_macro_names/3, []).
+    IgnoreModules = maps:get(ignore, RuleConfig, []),
+    Regex = maps:find(regex, RuleConfig),
```
Comment (elbrujohalcon) [inline src/elvis_style.erl:261]: What if we just use a default regex, like `[A-Z][A-Z_0-9]+`?

File: src/elvis_style.erl (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply src/elvis_style.erl:883]: I've never used the AST, but could try to check it out.

File: src/elvis_style.erl (hunk 7 above)
Comment (paulo-ferraz-oliveira) [reply src/elvis_style.erl:261]: That would break those cases where `to_upper` was valid (e.g. 'A-Z', even if ugly), but if it's something you prefer we can go with that.

Comment (elbrujohalcon) [reply src/elvis_style.erl:261]: Yeah, that was the original intention anyway. Let's break ugly macro names! 💣

File: src/elvis_style.erl (hunk 1 above)
Comment (elbrujohalcon) [reply src/elvis_style.erl:883]: There are several other rules using it, like the one for atoms you created in the other PR ;)

Comment (paulo-ferraz-oliveira) [reply src/elvis_style.erl:883]: Oh, you mean that `katana` stuff.

File: src/elvis_style.erl (hunk 7 above)
Comment (paulo-ferraz-oliveira) [reply src/elvis_style.erl]: Done in eca94aa.

File: src/elvis_style.erl [hunk 8]
```diff
@@ -37,9 +37,6 @@
 -define(NO_TRAILING_WHITESPACE_MSG,
         "Line ~b has ~b trailing whitespace characters.").
 
--define(INVALID_MACRO_NAME_MSG,
```
Comment (paulo-ferraz-oliveira) [inline src/elvis_style.erl:40]: This no longer makes sense, since we now have a regexp-based approach.

File: test/style_SUITE.erl [hunk 9]
```diff
@@ -246,7 +246,7 @@ verify_macro_names_rule(_Config) ->
     File = "fail_macro_names.erl",
     {ok, Path} = elvis_test_utils:find_file(SrcDirs, File),
 
-    [_, _, _] = elvis_style:macro_names(ElvisConfig, Path, #{}),
+    [_, _, _, _] = elvis_style:macro_names(ElvisConfig, Path, #{}),
```
Comment (paulo-ferraz-oliveira) [inline test/style_SUITE.erl:249]: A new one started to fail, since now were not simply considering `to_upper`.

File: src/elvis_style.erl (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply src/elvis_style.erl]: Done in 73f125c. I'll open a new discussion there, since there are details I'm not comfortable with.

File: src/elvis_style.erl [hunk 10]
```diff
@@ -874,30 +874,45 @@ check_no_trailing_whitespace(Line, Num, RuleConfig) ->
...
+                  Info = [MacroNameOriginal, Line, Regexp],
+                  Result = elvis_result:new(item, Msg, Info),
+                  ResultsIn ++ [Result];
+              {match, _Captured} ->
+                  ResultsIn
+          end,
+    check_macro_names(Regexp, RemainingMacroNodes, ResultsOut).
+
+is_macro_define_node(MaybeMacro) ->
+    case ktn_code:type(MaybeMacro) of
```
Comment (paulo-ferraz-oliveira) [inline src/elvis_style.erl:896]: Isn't there a better way to do this? I looked at what `katana` produced and basically got stuff from there, but I feel there's a better way.

File: src/elvis_style.erl [hunk 11]
```diff
@@ -874,30 +874,45 @@ check_no_trailing_whitespace(Line, Num, RuleConfig) ->
...
-    no_result;
-check_macro_name_regex(nomatch = _ReRunRes, {MacroName, LineNum, ConfigRegex}) ->
-    Msg = ?INVALID_MACRO_NAME_REGEX_MSG,
-    Info = [MacroName, LineNum, ConfigRegex],
-    Result = elvis_result:new(item, Msg, Info, LineNum),
-    {ok, Result}.
+macro_name_from_node(MacroNode) ->
+    MacroNodeValue = ktn_code:attr(value, MacroNode),
+    MacroAsAtom
+        = case lists:keyfind(_KeyVar = var, _N = 1, MacroNodeValue) of
```
Comment (paulo-ferraz-oliveira) [inline src/elvis_style.erl:906]: Also here, I feel there must be a better way to do this. Initially, I targeted type `macro` but that actually shows up when a macro is used, not defined, which is what we're interested in.

I later used `'ATOM'` in the macro definition, which provoked an error, since
```erlang
-define(MACRO, 2).
```
produces different `katana` structures from
```erlang
-define('MACRO', 2).
```
(first one is a `var`, second one is an `atom`).

Also, I didn't want to deep-dive into `katana`, nor could I find up-to-date information on the API.

**Comment by paulo-ferraz-oliveira** [comment]
Anyway, this is working (ups, tests are failing), and I'll want to `elvis` `elvis_core` since there's some stuff needing fixes there, too.

File: elvis.config [hunk 12]
```diff
@@ -6,7 +6,8 @@
      [#{dirs => ["src"],
         filter => "*.erl",
         rules => [{elvis_style, invalid_dynamic_call, #{ignore => [elvis_core]}},
-                  {elvis_style, dont_repeat_yourself, #{min_complexity => 20}}],
+                  {elvis_style, dont_repeat_yourself, #{min_complexity => 20}},
+                  {elvis_style, no_debug_call, #{ignore => [elvis_result, elvis_utils]}}],
```
Comment (paulo-ferraz-oliveira) [inline elvis.config:10]: Not too sure about this one, but there be `io:format` inside.

**Comment by paulo-ferraz-oliveira** [comment]
> I'll want to elvis-lint elvis_core

d50fbb2

File: elvis.config (hunk 12 above)
Comment (elbrujohalcon) [reply elvis.config:10]: Yeah, `elvis` is a CLI… it *has* to have `io:format` somewhere.

File: src/elvis_style.erl (hunk 10 above)
Comment (elbrujohalcon) [reply src/elvis_style.erl:899]: I think this is _the right way_:tm:.
Macros are very hard to parse. I learned that while implementing the formatter 🤦

File: src/elvis_style.erl (hunk 11 above)
Comment (elbrujohalcon) [reply src/elvis_style.erl:909]: It's very very ugly… but I think this is the only way.
The alternative would be to use `ktn_dodger`, as we do in the formatter, but that comes with another huge pile of nightmares.

**Review by elbrujohalcon — APPROVED** [review]
I think this is good to go.

**Comment by elbrujohalcon** [comment]
@paulo-ferraz-oliveira once you fix the tests, let me know and I can merge this.

File: elvis.config (hunk 12 above)
Comment (paulo-ferraz-oliveira) [reply elvis.config:10]: If only there were an `allow_debug_call` :D to which we would add non-ignored functions :D

**Comment by paulo-ferraz-oliveira** [comment]
What about the fact that `katana-code`'s `-spec(_).` is incomplete/wrong. Should we care about that (in the context of another repo, ofc)?

**Comment by paulo-ferraz-oliveira** [comment]
> @paulo-ferraz-oliveira once you fix the tests, let me know and I can merge this.

Pushed c414b1f. I don't exactly know what `katana-code`'s main goals are, but the new addition of code (about getting a macro's name) could well live inside it, right? ... since we're messing, in `elvis_core`'s context with `katana-code`'s internal structures.

**Comment by elbrujohalcon** [comment]
@jfacorro ☝️That's your territory :P

File: src/elvis_style.erl [hunk 13]
```diff
@@ -860,24 +876,49 @@ check_no_trailing_whitespace(Line, Num, RuleConfig) ->
...
-        {match, [MacroName]} ->
-            case string:to_upper(MacroName) of
-                MacroName ->
-                    no_result;
-                _ ->
-                    Msg = ?INVALID_MACRO_NAME_MSG,
-                    Result = elvis_result:new(item, Msg, [MacroName, Num], Num),
-                    {ok, Result}
-            end
+check_macro_names(_Regexp, [] = _MacroNodes, ResultsIn) ->
```
Comment (jfacorro) [inline src/elvis_style.erl:879]: Why the removal of the spec? 😄

**Comment by jfacorro** [comment]
> I don't exactly know what katana-code's main goals are, but the new addition of code (about getting a macro's name) could well live inside it, right? ... since we're messing, in elvis_core's context with katana-code's internal structures.

I couldn't find in this PR's code where we are manipulating internal katana-code data structures. As far as I could see we are using `ktn_code:attr/2` for getting values.

File: src/elvis_style.erl (hunk 13 above)
Comment (paulo-ferraz-oliveira) [reply src/elvis_style.erl:879]: You're right, this shouldn't have happened. Since I rewrote the function I accidentally took it out. In my lib.s I'm usually as strict as `missing_spec_all`. Shall I re-add it?

**Comment by paulo-ferraz-oliveira** [comment]
> > I don't exactly know what katana-code's main goals are, but the new addition of code (about getting a macro's name) could well live inside it, right? ... since we're messing, in elvis_core's context with katana-code's internal structures.
> 
> I couldn't find in this PR's code where we are manipulating internal katana-code data structures. As far as I could see we are using `ktn_code:attr/2` for getting values.

What I meant was "I'm handling a `katana-code` generated structure that can eventually change (https://github.com/inaka/elvis_core/pull/131/files#diff-9d4ce82551a31e44e1d37fb5c5966363R913) (and the code that does it, in the scope of `elvis` might surely be incomplete), especially since its `spec` specifies the return as `term()`..." 😄. I had proposed that this code be moved to `katana-code` since knowing about the internals shouldn't be required at this level, I guess. (I usually prefer opaque data structures that I "query" the producer for information about). No biggie, if we want it elsewhere, at least in the context of `elvis` it's now a function by itself.
